flusher.go

```diff
@@ -19,107 +20,70 @@ import (
-
-	percentiles := s.HistogramPercentiles
-
-	tempMetrics, ms := s.tallyMetrics(percentiles)
-
-	// the global veneur instance is also responsible for reporting the sets
-	// and global counters
-	ms.totalLength += ms.totalSets
-	ms.totalLength += ms.totalGlobalCounters
-
-	finalMetrics := s.generateInterMetrics(span.Attach(ctx), percentiles, tempMetrics, ms)
-
-	s.reportMetricsFlushCounts(ms)
-
-	s.reportGlobalMetricsFlushCounts(ms)
-
-	go func() {
-		for _, p := range s.getPlugins() {
-			start := time.Now()
-			err := p.Flush(span.Attach(ctx), finalMetrics)
-			s.Statsd.TimeInMilliseconds(fmt.Sprintf("flush.plugins.%s.total_duration_ns", p.Name()), float64(time.Since(start).Nanoseconds()), []string{"part:post"}, 1.0)
-			if err != nil {
-				countName := fmt.Sprintf("flush.plugins.%s.error_total", p.Name())
-				s.Statsd.Count(countName, 1, []string{}, 1.0)
-			}
-			s.Statsd.Gauge(fmt.Sprintf("flush.plugins.%s.post_metrics_total", p.Name()), float64(len(finalMetrics)), nil, 1.0)
-		}
-	}()
-
-	// TODO Don't hardcode this
-	s.metricSinks[0].Flush(span.Attach(ctx), finalMetrics)
-}
-
-// FlushLocal takes the slices of metrics, combines then and marshals them to json
-// for posting to Datadog.
-func (s *Server) FlushLocal(ctx context.Context) {
-	span, _ := trace.StartSpanFromContext(ctx, "")
-	defer span.Finish()
-
-	events, checks := s.EventWorker.Flush()
-	s.Statsd.Count("worker.checks_flushed_total", int64(len(checks)), nil, 1.0)
+	// TODO Concurrency
+	for _, sink := range s.metricSinks {
+		sink.FlushEventsChecks(span.Attach(ctx), events, checks)
+	}
 
-	go s.metricSinks[0].FlushEventsChecks(span.Attach(ctx), events, checks) // we can do all of this separately
-	go s.flushTraces(span.Attach(ctx))
+	go s.flushTraces(span.Attach(ctx)) // this too!
 
 	// don't publish percentiles if we're a local veneur; that's the global
 	// veneur's job
 	var percentiles []float64
+	var finalMetrics []samplers.InterMetric
+
+	if !s.IsLocal() {
+		percentiles = s.HistogramPercentiles
+	}
 
 	tempMetrics, ms := s.tallyMetrics(percentiles)
 
-	finalMetrics := s.generateInterMetrics(span.Attach(ctx), percentiles, tempMetrics, ms)
+	finalMetrics = s.generateInterMetrics(span.Attach(ctx), percentiles, tempMetrics, ms)
 
 	s.reportMetricsFlushCounts(ms)
 
-	// we don't report totalHistograms, totalSets, or totalTimers for local veneur instances
-
-	// we cannot do this until we're done using tempMetrics within this function,
-	// since not everything in tempMetrics is safe for sharing
-	go s.flushForward(span.Attach(ctx), tempMetrics)
+	if s.IsLocal() {
+		go s.flushForward(span.Attach(ctx), tempMetrics)
+	} else {
+		s.reportGlobalMetricsFlushCounts(ms)
+	}
 
 	// If there's nothing to flush, don't bother calling the plugins and stuff.
 	if len(finalMetrics) == 0 {
 		return
 	}
 
+	wg := sync.WaitGroup{}
+	for _, sink := range s.metricSinks {
+		wg.Add(1)
+		go func(ms metricSink) {
+			err := ms.Flush(span.Attach(ctx), finalMetrics)
```

Comment:
Now that this is happening asynchronously, we'll want to make doubly-sure that our interface definition specifies that the `Flush` method is not allowed to mutate the slice it receives.

Reply:
Definitely! How do we do that? :)

Reply:
I'd add it to the interface definition in `metric_sink.go`.